### PR TITLE
add load base split config and controller

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

###  What have you changed?

It's a part of #7400, it will be reviewed firstly.

###  What is the type of changes?

- New feature 

###  How is the PR tested?

- Unit test
- Integration test

### CODE REVIEW

File: components/raftstore/src/store/worker/split_config.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+}
+
+impl std::ops::Deref for SplitConfigManager {
+    type Target = Arc<VersionTrack<SplitConfig>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_config.rs:92]: need an empty line

File: components/raftstore/src/store/worker/split_controller.rs [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+}
+
+// It will return prefix sum of iter. `read` is a function to be used to read data from iter.
+fn prefix_sum<F, T>(iter: Iter<T>, read: F) -> Vec<usize>
+    where
+        F: Fn(&T) -> usize,
+{
+    let mut pre_sum = vec![];
+    for item in iter {
+        let last = match pre_sum.last() {
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:54]: use a sum variable will be easier?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+            sample_num: DEFAULT_SAMPLE_NUM,
+            region_infos: HashMap::default(),
+        }
+    }
+
+    pub fn add(&mut self, region_id: u64, peer: &Peer, key_range: KeyRange) {
+        self.batch_add(region_id, peer, vec![key_range]);
+    }
+
+    pub fn batch_add(&mut self, region_id: u64, peer: &Peer, key_ranges: Vec<KeyRange>) {
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:277]: Do we need both `add` and `batch_add`?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+            let order_end = if key_range.end_key.is_empty() {
+                Ordering::Less
+            } else {
+                sample.key.cmp(&key_range.end_key)
+            };
+
+            if order_start == Ordering::Greater && order_end == Ordering::Less {
+                sample.contained += 1;
+            } else {
+                if order_start != Ordering::Greater {
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:214]: How about directly using else if?

File: components/raftstore/src/store/worker/split_config.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+use configuration::{rollback_or, ConfigChange, ConfigManager, Configuration, RollbackCollector};
+use std::sync::Arc;
+use tikv_util::config::VersionTrack;
+
+const DEFAULT_DETECT_TIMES: u64 = 10;
+const DEFAULT_SAMPLE_THRESHOLD: i32 = 100;
+pub(crate) const DEFAULT_SAMPLE_NUM: usize = 20;
+const DEFAULT_QPS_THRESHOLD: usize = 1500;
+const DEFAULT_SPLIT_BALANCE_SCORE: f64 = 0.25;
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_config.rs:11]: How about adding some comments to explain these two parameters?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+        let interval = Duration::from_secs(self.cfg.detect_times * 2);
+        self.recorders
+            .retain(|_, recorder| recorder.create_time.elapsed().unwrap() < interval);
+    }
+
+    pub fn refresh_cfg(&mut self) {
+        if let Some(incoming) = self.cfg_tracker.any_new() {
+            self.cfg = incoming.clone();
+        }
+    }
```
Comment (djshow832) [inline components/raftstore/src/store/worker/split_controller.rs:377]: There' no lock here, are these two functions called in the same thread with `flush`?

File: components/raftstore/src/store/worker/split_controller.rs (hunk 3 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:277]: they will be used to `get` or `batch_get`

File: components/raftstore/src/store/worker/split_controller.rs (hunk 6 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:377]: Yes, the same thread

**Comment by lhy1024** [comment]
@rleungx @nolouch PTAL

File: components/raftstore/src/store/worker/split_controller.rs [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+
+    pub fn default() -> AutoSplitController {
+        AutoSplitController::new(SplitConfigManager::default())
+    }
+
+    pub fn flush(&mut self, others: Vec<QpsStats>) -> (Vec<usize>, Vec<SplitInfo>) {
+        let mut split_infos = Vec::default();
+        let mut top = BinaryHeap::with_capacity(TOP_N as usize);
+
+        let mut region_info_mat = HashMap::default();
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:307]: region_info_mat？

File: components/raftstore/src/store/worker/split_controller.rs [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+    }
+
+    #[bench]
+    fn qps_add(b: &mut test::Bencher) {
+        let mut qps_stats = default_qps_stats();
+        b.iter(|| {
+            qps_stats.add(0, &Peer::default(), build_key_range(b"a", b"b"));
+        });
+    }
+}
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:537]: new line

File: components/raftstore/src/store/worker/split_controller.rs [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+        for key_range in key_ranges {
+            if self.key_ranges.len() < self.sample_num {
+                self.key_ranges.push(key_range);
+            } else {
+                let i = rand::thread_rng().gen_range(0, self.qps) as usize;
+                if i < self.sample_num {
+                    self.key_ranges[i] = key_range;
+                }
+            }
+            self.qps += 1;
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:130]: Why `qps += 1` ?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+        sum += read(&item);
+        pre_sum.push(sum);
+    }
+    pre_sum
+}
+
+// It will return sample_num numbers by sample from lists.
+// The list in the lists has the length of N1, N2, N3 ... Np ... NP in turn.
+// Their prefix sum is pre_sum and we can get mut list from lists by get_mut.
+// Take a random number d from [1, N]. If d <N1, select a data in the first list with a equal probability without replacement;
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:64]:
```suggestion
// Take a random number d from [1, N]. If d < N1, select a data in the first list with an equal probability without replacement;
```

File: components/raftstore/src/store/worker/split_controller.rs [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+
+            let order_end = if key_range.end_key.is_empty() {
+                Ordering::Less
+            } else {
+                sample.key.cmp(&key_range.end_key)
+            };
+
+            if order_start == Ordering::Greater && order_end == Ordering::Less {
+                sample.contained += 1;
+            } else if order_start != Ordering::Greater {
```
Comment (nolouch) [inline components/raftstore/src/store/worker/split_controller.rs:211]: How about  start `Ordering::Less` and  end `Ordering:: Greater`?
```
key_range:          |_____________|
smaple:       |_________________________|
```

File: components/raftstore/src/store/worker/split_controller.rs [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+        }
+    }
+
+    fn is_ready(&self) -> bool {
+        self.times >= self.detect_num
+    }
+
+    fn collect(&mut self, config: &SplitConfig) -> Vec<u8> {
+        let pre_sum = prefix_sum(self.key_ranges.iter(), Vec::len);
+        let key_ranges = self.key_ranges.clone();
```
Comment (nolouch) [inline components/raftstore/src/store/worker/split_controller.rs:177]: Can this clone be avoided?

Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:177]: No, we get samples with equal probability without replacement, so we need to delete in line87

File: components/raftstore/src/store/worker/split_controller.rs (hunk 9 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:130]: `self.qps += key_ranges.len()` is better.

File: components/raftstore/src/store/worker/split_controller.rs (hunk 7 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:307]: `region_infos_map` maybe better.

**Comment by nolouch** [comment]
PTAL @rleungx

**Pull request merged**